### ISSUE TITLE
YOLOv5 Model Deployment: Unexpected Output Size, Normalization Issues, and Misaligned Bounding Boxes

### ISSUE THREAD

**Issue by mezgebbk** [ultralytics/hub#497]
### Search before asking

- [X] I have searched the HUB [issues](https://github.com/ultralytics/hub/issues) and found no similar bug report.

### HUB Component

Inference

### Bug

I trained a customized YOLOv5 model on specific data for a detection problem using Ultralytics HuB. After exporting and downloading the model as torchscript.pt, I encountered an issue during deployment in Android Studio for testing. The following lines of code were used for performing inference:
         Bitmap resizedBitmap = Bitmap.createScaledBitmap(mBitmap, 640, 640, true);
         final Tensor inputTensor = TensorImageUtils.bitmapToFloat32Tensor(resizedBitmap,
                TensorImageUtils.TORCHVISION_NORM_MEAN_RGB, TensorImageUtils.TORCHVISION_NORM_STD_RGB, 
                MemoryFormat.CHANNELS_LAST);
        final Tensor outputTensor = mModule.forward(IValue.from(inputTensor)).toTensor();
        final float[] outputs = outputTensor.getDataAsFloatArray();

The input tensor size is [1, 3, 640, 640], and the output tensor size obtained was [1, 7, 8400]. Shouldn't the expected output size be 25200? Additionally, as shown in the attached screenshot, the output values are not normalized. Wasn't I supposed to obtain a normalized output after performing inference on an image input? Moreover, the bounding boxes seem misaligned with the sample, resulting in inaccurate predictions.

If you require any further information or code, please feel free to ask. Thank you for your assistance!

![input output tensor](https://github.com/ultralytics/hub/assets/89604316/33b67952-1117-433b-8d5f-95c44b38bd39)
![outpur result](https://github.com/ultralytics/hub/assets/89604316/90ac5508-63b7-4cad-9854-244435fddb83)

**Comment by UltralyticsAssistant** [comment]
@mezgebbk hello! Thank you for reaching out with your issue.

Regarding the output tensor size of [1, 7, 8400], this may reflect your model's configured number of anchors, classes, and grid size. Ensure your model's configuration matches the expected layout for YOLOv5 where the output tensor typically represents `[1, number of anchors * (classes + 5), grid size]`. If you have different configurations, the output tensor size might indeed differ.

Output values not appearing normalized can result from several factors, including but not limited to, post-processing steps applied or omitted in the deployment code, or discrepancies in the exported model. Normalization typically occurs before model inference; output values usually represent raw predictions requiring further processing, like applying an activation function and thresholding.

Misaligned bounding boxes suggest that the post-processing steps to convert raw outputs into actual coordinates may not be correctly implemented or the model weights aren't appropriately interpreted by the inference code. Review your code to ensure that all necessary transformation steps, such as scaling the bounding box coordinates back to the original image size and applying the correct objectness and class confidence thresholds, are correctly performed.

I highly recommend reviewing the [Ultralytics HUB Docs](https://docs.ultralytics.com/hub) for further guidance on best practices for model deployment and inference, including post-processing steps.

Let us know if there's anything specific that isn't covered in the documentation or if you need more clarification on this topic. Thank you for your commitment to the YOLO community, and we're here to help you solve these issues. 🚀

**Comment by mezgebbk** [comment]
Thank you for your response. The training was conducted using the Ultralytics Hub (Image Size 640), and subsequently, I exported the model, downloading it as 'model.torchscipt.pt'. Before conducting inference, I ensured that my input image was normalized using the same scaling as employed during training, 
MEAN_RGB = new float[] {0.485f, 0.456f, 0.406f}
STD_RGB = new float[] {0.229f, 0.224f, 0.225f}
and below is the normalized input tensor. However, despite these normalized input tensor, the output obtained after performing inference appears to be non-normalized. The inference is performed using 

        mModule = Module.load(assetFilePath(MainActivity.this, "model.torchscipt.pt"))
        final Tensor outputTensor = mModule.forward(IValue.from(inputTensor)).toTensor();
        final float[] outputs = outputTensor.getDataAsFloatArray();

![input output tensor](https://github.com/ultralytics/hub/assets/89604316/d430c82c-4f2b-4f33-8a4a-7c550a99af92)
![Ultralytics](https://github.com/ultralytics/hub/assets/89604316/a235c62c-a33c-4f28-adf7-43dfff69c4ad)

**Comment by UltralyticsAssistant** [comment]
Thank you for the additional information, @mezgebbk.

It's great to hear that you're normalizing your input images correctly with the same mean and standard deviation used during training. That practice is essential for consistent model performance.

As for the output tensor appearing non-normalized, that is generally expected behavior. The raw outputs of a model are typically logits, confidence scores, and/or bounding box coordinates, which are not normalized in the same way as input images. These outputs require post-processing to interpret; this may involve applying a sigmoid or softmax function to probabilities, thresholding to filter out low-confidence predictions, and transforming the bounding box coordinates based on the image's size.

If the bounding boxes remain misaligned even after ensuring the input normalizations and configurations match the training settings, then it might be necessary to:
1. Confirm your post-processing steps align with the expectations for YOLOv5 outputs, including any activation functions and transforming the output grid coordinates to image space.
2. Double-check that the dimensions and scales of the bounding boxes are being correctly calculated and adjusted for the original image resolution.
3. Verify that the model weights loaded in Android match the ones used and validated after training. Sometimes, issues during the export or load phase can cause discrepancies.

Remember to refer to the Ultralytics HUB Docs, especially sections detailing post-processing steps and deployment-to-production workflow, which should provide further insights into how to handle outputs correctly.

If your investigation indicates the discrepancy might be due to an implementation detail not covered by the documentation or a potential issue with the Ultralytics HUB export feature, please provide us with such details, and we can look into it more deeply.

We are here to support you, so keep us posted on your progress! 🌟
